File: src/main/java/com/und/eventapi/model/Event.java
package com.und.eventapi.model;

import com.fasterxml.jackson.annotation.JsonProperty;
import io.swagger.annotations.ApiModelProperty;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.mapping.Document;

import javax.validation.constraints.NotNull;
import java.util.HashMap;

/**
 * Created by shiv on 21/07/17.
 */
//@Document(collection = "#{tenantId}_event")
public class Event {

    @Id
    private String id;

    private Long userid = null;

    private Boolean isanonymus = null;

    private String username = null;

    private String firstName = null;

    private String lastName = null;

    private String email = null;

    private String phone = null;

    private String time = null;


    private String name;

    private HashMap<String, String> map = null;

    public Event() {
    }

    public Event(String name) {
        this.name = name;
    }

    @ApiModelProperty(required = true, value = "unique event name")
    @NotNull
    public String getName() {
        return name;
    }


    public void setName(String name) {
        this.name = name;
    }


    /**
     * userid to track
     *
     * @return userid
     **/
    @ApiModelProperty(required = true, value = "userid to track")
    @NotNull
    public Long getUserid() {
        return userid;
    }

    public void setUserid(Long userid) {
        this.userid = userid;
    }

    public Event isanonymus(Boolean isanonymus) {
        this.isanonymus = isanonymus;
        return this;
    }

    /**
     * if this property is true user details are tracked only for this session. all other user details are ignored.
     *
     * @return isanonymus
     **/
    @ApiModelProperty(value = "if this property is true user details are tracked only for this session. all other user details are ignored. ")
    public Boolean getIsanonymus() {
        return isanonymus;
    }

    public void setIsanonymus(Boolean isanonymus) {
        this.isanonymus = isanonymus;
    }

    public Event username(String username) {
        this.username = username;
        return this;
    }

    /**
     * Get username
     *
     * @return username
     **/
    @ApiModelProperty(value = "")
    public String getUsername() {
        return username;
    }

    public void setUsername(String username) {
        this.username = username;
    }

    public Event firstName(String firstName) {
        this.firstName = firstName;
        return this;
    }

    /**
     * Get firstName
     *
     * @return firstName
     **/
    @ApiModelProperty(value = "")
    public String getFirstName() {
        return firstName;
    }

    public void setFirstName(String firstName) {
        this.firstName = firstName;
    }

    public Event lastName(String lastName) {
        this.lastName = lastName;
        return this;
    }

    /**
     * Get lastName
     *
     * @return lastName
     **/
    @ApiModelProperty(value = "")
    public String getLastName() {
        return lastName;
    }

    public void setLastName(String lastName) {
        this.lastName = lastName;
    }

    public Event email(String email) {
        this.email = email;
        return this;
    }

    /**
     * Get email
     *
     * @return email
     **/
    @ApiModelProperty(value = "")
    public String getEmail() {
        return email;
    }

    public void setEmail(String email) {
        this.email = email;
    }

    public Event phone(String phone) {
        this.phone = phone;
        return this;
    }

    /**
     * Get phone
     *
     * @return phone
     **/
    @ApiModelProperty(value = "")
    public String getPhone() {
        return phone;
    }

    public void setPhone(String phone) {
        this.phone = phone;
    }

    public Event time(String time) {
        this.time = time;
        return this;
    }

    /**
     * Get time
     *
     * @return time
     **/
    @ApiModelProperty(value = "")
    public String getTime() {
        return time;
    }

    public void setTime(String time) {
        this.time = time;
    }

    @ApiModelProperty(value = "")
    public HashMap<String, String> getMap() {
        return map;
    }

    public void setMap(HashMap<String, String> map) {
        this.map = map;
    }


}

File: src/main/kotlin/com/und/eventapi/model/Event.kt
package com.und.eventapi.model

import com.und.eventapi.validation.ValidateDate
import com.und.model.mongo.LineItem
import org.springframework.format.annotation.DateTimeFormat
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.ZoneOffset
import java.util.*
import javax.validation.constraints.*

/**
 * Created by shiv on 21/07/17.
 */

//FIXME handle validation for other fields and different types, e.g. @see
//@ValidateDate
open class Event {

    @NotBlank(message = "{event.name.empty}")
    lateinit var name: String

    @NotNull(message="{event.clientid.null}")
    var clientId: Int = -1

    var identity: Identity = Identity()
    var creationTime: Long = LocalDateTime.now().toInstant(ZoneOffset.UTC).toEpochMilli()

    @Pattern(regexp="'\\b((25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)(\\.|\$)){4}\\b'",
            message="{event.ip.invalid}")
    var ipAddress: String? = null

    @Size(min = 3, max = 25, message = "{event.city.invalidSize}")
    @Pattern(regexp = "^[A-Za-z][a-zA-Z\\s]*$",message = "{event.city.invalidCharacters}")
    var city: String? = null

    @Size(min = 4, max = 25, message = "{event.state.invalidSize}")
    @Pattern(regexp = "^[A-Za-z][a-zA-Z\\s]*$",message = "{event.state.invalidCharacters}")
    var state: String? = null

    @Size(min = 4, max = 25, message = "{event.country.invalidSize}")
    @Pattern(regexp = "^[A-Za-z][a-zA-Z\\s]*$",message = "{event.country.invalidCharacters}")
    var country: String? = null

    @Pattern(regexp = "^[-+]?([1-8]?\\d(\\.\\d+)?|90(\\.0+)?),\\s*[-+]?(180(\\.0+)?|((1[0-7]\\d)|([1-9]?\\d))(\\.\\d+)?)\$",
            message="{event.latitude.invalid}")
    var latitude: String? = null

    @Pattern(regexp = "^[-+]?([1-8]?\\d(\\.\\d+)?|90(\\.0+)?),\\s*[-+]?(180(\\.0+)?|((1[0-7]\\d)|([1-9]?\\d))(\\.\\d+)?)\$",
            message="{event.longitude.invalid}")
    var longitude: String? = null

    var agentString: String? = null
    var userIdentified: Boolean = false
    var lineItem: MutableList<LineItem> = mutableListOf()
    var attributes: HashMap<String, Any> = hashMapOf()

    @DateTimeFormat(pattern = "yyyy-MM-dd")
    var startDate= LocalDate.now()

    @DateTimeFormat(pattern = "yyyy-MM-dd")
    var endDate=LocalDate.of(2017, 1, 13)

}

data class Identity(
        //unique id assigned to a device, should always remain fixed, create new if not found
        var deviceId: String = "",
        //if userId is not found assign a new session id, handle change if user login changes, logouts etc
        var sessionId: String = "",
        // id of event user, this id is assigned when a user profile is identified.
        var userId: String? = null,
        var clientId: Int? = -1
)

class date : Event() {
    //TODO
}

/* {
    var eventUser: EventUser = EventUser()
}
*/





